docs/migration.md

```diff
@@ -60,36 +60,56 @@ This document offers guidance on migrating from Prometheus 2.x to Prometheus 3.0
 
-- The `.` pattern in regular expressions in PromQL matches newline characters. 
-  With this change a regular expressions like `.*` matches strings that include 
-  `\n`. This applies to matchers in queries and relabel configs.
-  - For example, the following regular expressions now match the accompanying
-  strings, whereas in Prometheus v2 these combinations didn't match.
-      - `.*` additionally matches `foo\n` and `Foo\nBar`
-      - `foo.?bar` additionally matches `foo\nbar`
-      - `foo.+bar` additionally matches `foo\nbar`
-  - If you want Prometheus v3 to behave like v2, you will have to change your
-  regular expressions by replacing all `.` patterns with `[^\n]`, e.g.
-  `foo[^\n]*`.
-- Lookback and range selectors are left open and right closed (previously left 
-  closed and right closed). This change affects queries when the evaluation time 
-  perfectly aligns with the sample timestamps. For example assume querying a 
-  timeseries with evenly spaced samples exactly 1 minute apart. Before Prometheus
-  v3, a range query with `5m` would usually return 5 samples. But if the query
-  evaluation aligns perfectly with a scrape, it would return 6 samples. In 
-  Prometheus v3 queries like this will always return 5 samples.  
-  This change has likely few effects for everyday use, except for some subquery
-  use cases.  
-  Query front-ends that align queries usually align subqueries to multiples of
-  the step size. These subqueries will likely be affected.  
-  Tests are more likely to affected. To fix those either adjust the expected 
-  number of samples or extend the range by less than one sample interval.
-- The `holt_winters` function has been renamed to `double_exponential_smoothing` 
-  and is now guarded by the `promql-experimental-functions` feature flag.
-  If you want to keep using `holt_winters`, you have to do both of these things:
-    - Rename `holt_winters` to `double_exponential_smoothing` in your queries.
-    - Pass `--enable-feature=promql-experimental-functions` in your Prometheus 
-      CLI invocation.
+### Regexes match newlines
+
+The `.` pattern in regular expressions in PromQL matches newline characters.
+With this change a regular expressions like `.*` matches strings that include
+`\n`. This applies to matchers in queries and relabel configs.
+
+For example, the following regular expressions now match the accompanying
+strings, whereas in Prometheus v2 these combinations didn't match.
+    - `.*` additionally matches `foo\n` and `Foo\nBar`
+    - `foo.?bar` additionally matches `foo\nbar`
+    - `foo.+bar` additionally matches `foo\nbar`
+
+If you want Prometheus v3 to behave like v2, you will have to change your
+regular expressions by replacing all `.` patterns with `[^\n]`, e.g.
+`foo[^\n]*`.
+
+### Intervals return a predictable number of points
+
+Interval selectors are now more predictable than they were previously, always
+returning the same number of points regardless of query alignment.
```

Comment:
That's not true in general, not even for sub-queries. (If you say `[5m:2m]`, you might get two or three samples, depending on alignment.)

As said, I think starting with "predictable number of samples" is not a suitable starting point. It's just one specific outcome of this in cases like `[5m:1m]`.

Reply:
what do you think is the better way to describe the benefit of the new system?

Reply:
tried rewording

Reply:
The benefit of the new system is manifold. It depends on your specific use case what aspects are most important.

Therefore, I would indeed first state that ranges are left-open now. The rest follows.

For example, if I evaluate `foo[1m]` now and 1m ago, I would expect to get a completely distinct set of samples. However, that was not true in v2 if and only if there was a sample at precisely 1m ago. So the behavior is much more benign now, but saying that "the number of samples is predictable" now would be misleading. It was predictable even before, just that it needed more data to predict.

Reply:
tried again